### PR TITLE
Add dry-run execution to validate BigQuery imports

### PR DESCRIPTION
**Pull request by jgoizueta**
See #15383

This detects permission errors (or other errors like invalid SQL syntax, non-existent tables or columns, etc.) by performing a dry run execution of the query.

This does not fix the problem completely, since we cannot check Storage API in dry mode and we don't have a Ruby client anyway. If Storage API is used, silent errors due to missing permissions to use the API may still produce empty data results. 

TODO: check for empty imports and warn the user somehow.

### CODE REVIEW

**Comment by jgoizueta** [comment]
A dryrun is now always performed before executing the actual import and an error occurs if permissions are not ok, table does not exist etc. But the nice and informative error message from BQ API does not reach to the user when this happens during import (neither import API or UI). 
As a workaround for the time being I've added a `dryrun` endpoint to explicitly execute the dry run that returns the error information and the bytes to be processed, so that the UI can show this information to the user.

Pending:
- [ ] fix tests for new metadata
- [ ] document new endpoints
- [ ] unify dryrun result format (error vs no error)
- [ ] put dryrun errors in import state
- [ ] consider pagination for listing all tables

File: lib/carto/connector/providers/fdw/odbc/bigquery.rb [hunk 1]
```diff
@@ -68,11 +81,42 @@ def table_options
...
+      def table_query
+        project = @params[:project] || @params[:billing_project]
+        %{SELECT * FROM `#{project}.#{@params[:dataset]}.#{@params[:table]}`;}
+      end
+
+      def list_tables(limits: {})
+        tables = []
+        limit = limits[:max_listed_tables]
+        oauth_client = @sync_oauth&.get_service_datasource
+        if oauth_client
```
Comment (gonzaloriestra) [inline lib/carto/connector/providers/fdw/odbc/bigquery.rb]: To reduce one nested level, you could remove this conditional and instead do:
```
projects = oauth_client&.list_projects || []
projects.each...
```

File: lib/carto/connector/providers/fdw/odbc/bigquery.rb [hunk 2]
```diff
@@ -182,6 +226,19 @@ def fixed_odbc_attributes
           end
         end
 
+        unless @oauth_config['no_dry_run']
+          # Perform a dry-run of the query to catch errors (API permissions, SQL syntax, etc.)
+          # Note that the import may stil fail if using Storage API and needed permission is missing.
```
Comment (gonzaloriestra) [inline lib/carto/connector/providers/fdw/odbc/bigquery.rb]:
```suggestion
          # Note that the import may still fail if using Storage API and needed permission is missing.
```

File: lib/carto/connector/providers/fdw/odbc/bigquery.rb [hunk 3]
```diff
@@ -182,6 +226,19 @@ def fixed_odbc_attributes
...
+          sql = @params[:sql_query] || table_query
+          result = perform_dry_run(@params[:billing_project], sql)
```
Comment (gonzaloriestra) [inline lib/carto/connector/providers/fdw/odbc/bigquery.rb]: Why not save it directly into @dry_run_result?

File: lib/carto/connector/providers/fdw/odbc/bigquery.rb [hunk 4]
```diff
@@ -182,6 +226,19 @@ def fixed_odbc_attributes
...
+          @dry_run_result = result
+          if result[:error]
+            # TODO: avoid rescuing errors in dry_run? return our own exception here?
+            raise result[:client_error]
+          end
```
Comment (gonzaloriestra) [inline lib/carto/connector/providers/fdw/odbc/bigquery.rb]: The ruby way would be in one line:
`raise result[:client_error] if result[:error]`

File: services/datasources/lib/datasources/url/bigquery.rb [hunk 5]
```diff
@@ -214,6 +214,59 @@ def create_dataset(project_id, dataset_id, options)
           @bigquery_api.insert_dataset(project_id, dataset)
         end
 
+        def list_projects
+          projects = @bigquery_api.list_projects.projects
+          if projects
+            projects.map { |p| { id: p.id, friendly_name: p.friendly_name } }
+          else
+            []
```
Comment (gonzaloriestra) [inline services/datasources/lib/datasources/url/bigquery.rb]: It's also preferred to use guard clauses:
```
return [] unless projects
projects.map { |p| { id: p.id, friendly_name: p.friendly_name } }
```

**Comment by jgoizueta** [comment]
Regarding the new endpoints specs I had my doubts because it would be hard to make them work (even with some mocking) in CI. I'll add them for reference but leave most of the cases as _pending_.

**Pull request merged**